### DISCUSSION TITLE
How do I pass a value from a middleware to controllers?

### DISCUSSION THREAD

**Discussion by afzalIbnSH** [vercel/next.js#34914] [Help]
Hi,

Coming from the node.js-express world and working with ts-next.js for the first time, I've been scratching my head as to how I could pass a value from a middleware to the subsequent controllers.

My exact requirement is to carry out token-based authentication in the middleware and pass the retrieved user data to all the subsequent controllers.

In node.js-express, I would assign the details to req.user or add an attribute to res.locals. Neither works here for obvious reasons. What's the way to go here?, other than calling a function in every single controller.

**Comment by otterlistic** [comment]
[This](https://github.com/vercel/next.js/issues/30534) issue seems to refer to the same thing. IMO calling edge functions "middleware" is kinda misleading when the environment that a "middleware" and an API route executes in are completely different. You'd expect them to be one and the same because that's what middleware is in most contexts, something that just runs before your actual business logic.

Regardless, I'd really love to hear of solutions/workarounds myself.

There's currently also a very annoying lack of a standard routing solution when it comes to NextJS API routes IMO.
